addons/test_mail_sms/tests/test_sms_post.py

```diff
@@ -44,6 +44,23 @@ def test_message_sms_internals_resend_existingd(self):
             test_record._notify_thread_by_sms(messages, [{'id': self.partner_1.id, 'notif': 'sms'}], resend_existing=True)
         self.assertSMSNotification([{'partner': self.partner_1}], self._test_body, messages)
 
+    def test_message_sms_additonal_number_same_as_parter_number(self):
+        """ _message_sms ( which uses _notify_thread_by_sms) allows for specyfing additional number to send sms to
+            This test checks for situation where this additional number is the same as partner telephone number.
+            In that case sms shall NOT be sent twice."""
+        with self.with_user('employee'), self.mockSMSGateway():
+            test_record = self.env['mail.test.sms'].browse(self.test_record.id)
+            additional_number_same_as_partner_number = self.partner_1.mobile
+            subtype_id = self.env['ir.model.data']._xmlid_to_res_id('mail.mt_note')
+            test_record._message_sms(
+                body=self._test_body,
+                partner_ids=self.partner_1.ids,
+                subtype_id=subtype_id,
+                sms_numbers=[additional_number_same_as_partner_number],
+                number_field='mobile'
+            )
+        self.assertEqual(self.env['sms.sms'].sudo().search_count([]), 1, "Message got duplicated.")
```

Comment:
Don't you have tools available with the mock instead ? Here basically you depend on demo, existing data, ...

Reply:
Can you elaborate? I'm not sure what do you mean. I followed the concept of reusing as much as possible from the existing test setup.

Reply:
You have mock tools (like mockSMSGateway but you already use it) and custom asserts (like assertSMSNotification) available in common.py, allowing to be sure your test is reproducible. For example here, you do a search on sms.sms, but you could have some of them already existing in db when running the test and it would fail. Or it could be another one. Custom asserts normally check for sms created during the mock, so it would effectively check what happens during _message_sms.

Reply:
Thanks for elaboration. I checked available asserts in the **common.py**, but I don't see any particular assert suitable to my needs (just counting new sms), as most of them are designed to check for content. They also perform filtration for sms partner_id which is missing in the buggy scenario (so I can't use `_find_sms_sms` for example)

However to relay on `mockSMSGateway` I've adopted assert to count `_new_sms` (for the partner number) instead of searching `sms.sms` myself. Let me know if there's another way to improve this.

Reply:
Ok, good enough :)